lib/active_merchant/billing/gateways/trans_first_transaction_express.rb

```diff
@@ -0,0 +1,597 @@
+        "A8" => "Reserved for future Realtime use",
+        "B1" => "Surcharge amount not permitted on Visa cards or EBT Food Stamps/ Reserved for future Realtime use",
+        "B2" => "Surcharge amount not supported by debit network issuer/ Reserved for future Realtime use",
+        "C1" => "Unacceptable PIN",
+        "C2" => "PIN Change failed",
+        "C3" => "PIN Unblock failed",
+        "D1" => "MAC Error",
+        "E1" => "Prepay error",
+        "N1" => "Network Error within the TXP platform",
+        "N0" => "Force STIP/ Reserved for client-specific use (declined)",
+        "N3" => "Cash service not available/ Reserved for client-specific use (declined)",
+        "N4" => "Cash request exceeds Issuer limit/ Reserved for client-specific use (declined)",
+        "N5" => "Ineligible for re-submission/ Reserved for client-specific use (declined)",
+        "N7" => "Decline for CVV2 failure/ Reserved for client-specific use (declined)",
+        "N8" => "Transaction amount exceeds preauthorized approval amount/ Reserved for client-specific use (declined)",
+        "P0" => "Approved; PVID code is missing, invalid, or has expired",
+        "P1" => "Declined; PVID code is missing, invalid, or has expired/ Reserved for client-specific use (declined)",
+        "P2" => "Invalid biller Information/ Reserved for client-specific use (declined)/ Reserved for client-specific use (declined)",
+        "R0" => "The transaction was declined or returned, because the cardholder requested that payment of a specific recurring or installment payment transaction be stopped/ Reserved for client-specific use (declined)",
+        "R1" => "The transaction was declined or returned, because the cardholder requested that payment of all recurring or installment payment transactions for a specific merchant account be stopped/ Reserved for client-specific use (declined)",
+        "Q1" => "Card Authentication failed/ Reserved for client-specific use (declined)",
+        "XA" => "Forward to Issuer/ Reserved for client-specific use (declined)",
+        "XD" => "Forward to Issuer/ Reserved for client-specific use (declined)",
+      }
+
+      EXTENDED_RESPONSE_MESSAGES = {
+        "B40K" => "Declined Post – Credit linked to unextracted settle transaction"
+      }
+
+      TRANSACTION_CODES = {
+        authorize: 0,
+        void_authorization: 2,
+
+        purchase: 1,
+        capture: 3,
+        void_purchase: 6,
+        void_capture: 6,
+
+        refund: 4,
+        blind_credit: 5,
+        void_refund: 13,
+        void_credit: 13,
+
+        verify: 9,
+
+        wallet_sale: 14,
+      }
+
+      def initialize(options={})
+        requires!(options, :gateway_id, :reg_key)
+        super
+      end
+
+      def purchase(amount, payment_method, options={})
+        if credit_card? payment_method
+          request = build_xml_transaction_request do |doc|
+            add_transaction_code(doc, :purchase)
```

Comment:
The trouble there is that the transaction needs to be written deep into the XML document. If we build it all up it wouldn't be easy to insert the transaction code.